packages/java/maven-plugin/src/main/java/dev/hilla/maven/EngineGenerateMojo.java

```diff
@@ -0,0 +1,76 @@
+package dev.hilla.maven;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Objects;
+
+import dev.hilla.internal.EngineConfiguration;
+import dev.hilla.internal.EngineException;
+import dev.hilla.internal.EngineRunner;
+import dev.hilla.internal.GeneratorConfiguration;
+import dev.hilla.internal.GeneratorUnavailableException;
+import dev.hilla.internal.ParserConfiguration;
+
+import org.apache.maven.plugin.AbstractMojo;
+import org.apache.maven.plugins.annotations.Execute;
+import org.apache.maven.plugins.annotations.LifecyclePhase;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.maven.plugins.annotations.ResolutionScope;
+import org.apache.maven.project.MavenProject;
+
+/**
+ * Maven Plugin for Hilla. Handles parsing Java bytecode and generating
+ * TypeScript code from it.
+ */
+@Mojo(name = "generate", defaultPhase = LifecyclePhase.PROCESS_CLASSES, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME)
+@Execute(goal = "configure")
+public final class EngineGenerateMojo extends AbstractMojo {
+
+    @Parameter(readonly = true)
+    private final GeneratorConfiguration generator = new GeneratorConfiguration();
+
+    @Parameter(readonly = true)
+    private final ParserConfiguration parser = new ParserConfiguration();
+
+    @Parameter(defaultValue = "${project}", required = true, readonly = true)
+    private MavenProject project;
+
+    @Parameter(defaultValue = "${project.build.directory}", readonly = true)
+    private File buildDirectory;
+
+    // If set to false, the plugin will not fail if the generator is not
+    // available. This allows to run this goal just to save the configuration,
+    // without having to run 'npm install' to avoid an error
+    @Parameter(property = "hilla.failOnMissingGenerator", defaultValue = "true")
+    private boolean failOnMissingGenerator;
+
+    @Override
+    public void execute() throws EngineGenerateMojoException {
+        try {
+            var conf = Objects.requireNonNull(
+                EngineConfiguration.load(buildDirectory));
+            var classPath = conf.getClassPath();
+            var urls = new ArrayList<URL>(classPath.size());
+            for (var classPathItem : classPath) {
+                urls.add(new File(classPathItem).toURI().toURL());
+            }
+            var classLoader = new URLClassLoader(urls.toArray(URL[]::new));
```

Comment:
What if we provide an existing class loader here? like `getClass().getClassLoader()`? Wouldn't it work better and provide an uniform access for all classes loaded by Maven?

Reply:
To my knowledge, `getClass().getClassLoader()` would not work. According to the [Maven Class Loading article](https://cwiki.apache.org/confluence/display/MAVEN/Maven+3.x+Class+Loading):

> The class path of a plugin class loader is derived from the dependencies of the plugin as defined in the POM of the plugin artifact and optionally by user-defined plugin dependencies listed in the <plugin> declaration of the project that uses the plugin. The plugin class loader does neither contain the dependencies of the current project nor the compiled main/test classes of the project. Plugins that need to load classes from the compile/runtime/test class path of a project need to create a custom URLClassLoader in combination with the mojo annotation @requiresDependencyResolution.

Reply:
Hmm, ok, I see. In my case, using `getClass().getClassLoader()` in the `Parser` worked perfectly but I don't know if it is going to work with your changes.